File: spring-rest-configuration-autodiscover.java/src/main/java/demo/controller/ConfigurationController.java
package demo.controller;

import demo.domain.Property;
import demo.domain.PropertyHistory;
import demo.dto.Value;
import demo.service.ConfigurationHistoryService;
import demo.service.ConfigurationService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping(value = "api/configuration", produces = MediaType.APPLICATION_JSON_VALUE)
public class ConfigurationController {

    @Autowired
    private ConfigurationService configurationService;

    @Autowired
    private ConfigurationHistoryService configurationHistoryService;

    @RequestMapping(value = "", method = RequestMethod.GET)
    public ResponseEntity<List<Property>> findAllProperties() {
        return new ResponseEntity<>(configurationService.findAllProperties(), HttpStatus.OK);
    }

    @RequestMapping(value = "/name/{name}", method = RequestMethod.GET)
    public ResponseEntity<Property> findPropertyByName(
            @PathVariable String name) {
        Optional<Property> propertyOpt = configurationService.findPropertyByName(name);
        return propertyOpt
                .map(configuration -> new ResponseEntity<>(configuration, HttpStatus.OK))
                .orElseGet(() -> new ResponseEntity<>(new Property(), HttpStatus.NOT_FOUND));
    }

    @RequestMapping(value = "/name/{name}", method = RequestMethod.PUT, consumes = MediaType.APPLICATION_JSON_VALUE)
    public ResponseEntity<Property> updateProperty(
            @PathVariable String name,
            @Valid @RequestBody Value value) {
        return new ResponseEntity<>(configurationService.updateProperty(name, value.getValue()), HttpStatus.OK);
    }

    @RequestMapping(value = "/history", method = RequestMethod.GET)
    public List<PropertyHistory> findAllPropertyHistories() {
        return configurationHistoryService.findAllPropertyHistories();
    }
}

File: spring-rest-configuration-autodiscover.scala/src/main/scala/demo/controller/ConfigurationController.scala
package demo.controller

import javax.validation.Valid

import demo.domain.{Property, PropertyHistory}
import demo.dto.Value
import demo.service.{ConfigurationHistoryService, ConfigurationService}
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.{HttpStatus, MediaType, ResponseEntity}
import org.springframework.web.bind.annotation._

@RestController
@RequestMapping(value = Array("api/configuration"), produces = Array(MediaType.APPLICATION_JSON_VALUE))
class ConfigurationController {

  @Autowired
  private val configurationService: ConfigurationService = null

  @Autowired
  private val configurationHistoryService: ConfigurationHistoryService = null

  @RequestMapping(value = Array(""), method = Array(RequestMethod.GET))
  def findAllProperties(): ResponseEntity[java.util.List[Property]] = {
    new ResponseEntity(configurationService.findAllProperties(), HttpStatus.OK)
  }

  @RequestMapping(value = Array("/name/{name}"), method = Array(RequestMethod.GET))
  def findPropertyByName(
                          @PathVariable name: String): ResponseEntity[Property] = {
    val propertyOpt = configurationService.findPropertyByName(name)
    propertyOpt
      .map[ResponseEntity[Property]]({ property => new ResponseEntity(property, HttpStatus.OK) })
      .orElseGet { () => new ResponseEntity(new Property(), HttpStatus.NOT_FOUND) }
  }

  @RequestMapping(value = Array("/name/{name}"), method = Array(RequestMethod.PUT), consumes = Array(MediaType.APPLICATION_JSON_VALUE))
  def updateProperty(
                      @PathVariable name: String,
                      @Valid @RequestBody value: Value): ResponseEntity[Property] = {
    new ResponseEntity(configurationService.updateProperty(name, value.value), HttpStatus.OK)
  }

  @RequestMapping(value = Array("/history"), method = Array(RequestMethod.GET))
  def findAllPropertyHistories(): java.util.List[PropertyHistory] = {
    configurationHistoryService.findAllPropertyHistories()
  }
}
